### PR TITLE
Reorganize Reaction Ensemble

### LINKED ISSUE #4155: Separation of concerns in Reaction Ensemble
The RE needs to be moved to a submodule of the core and split in several files, one per RE class (along with their dependencies). Right now all classes are mixed together, which makes maintenance more difficult. For example:
* It is not immediately clear which utility classes are needed by each RE class. It's also unclear if RE classes depend on each other, or if they only depend on the parent class `ReactionAlgorithm`. Having separate files would help figure that information out by simply reading the `#include` statements.
* The public API of RE includes `particle_data.hpp` and `energy.hpp`, both of which shouldn't be exposed. They are actually only needed in the implementation of private methods of the `CollectiveVariable` classes and could easily be moved to the .cpp file (the overhead of not having them `inline` shouldn't matter much). `particle_data.hpp` in particular contains prototypes for particle communication and should only be included in .cpp files; in fact, the `Particle` structs were extracted from `particle_data.hpp` and stored in `Particle.hpp` in ESPResSo 4.2 to help us hide the communication logic and remove as many `#include <particle_data.hpp>` as possible (75ef537f9d656b, 78bf03deefd7145e).

The new layout could look like this:
```
src/core/reaction_ensemble/
                          |- CMakeLists.txt
                          |- utils.hpp
                          |- utils.cpp
                          |- ReactionAlgorithm.hpp
                          |- ReactionAlgorithm.cpp
                          |- ConstantpHEnsemble.hpp
                          |- ConstantpHEnsemble.cpp
                          |- WidomInsertion.hpp
                          |- WidomInsertion.cpp
                          |- ReactionEnsemble.hpp
                          |- ReactionEnsemble.cpp
                          |- WangLandauReactionEnsemble.hpp
                          |- WangLandauReactionEnsemble.cpp
```

Special files:
* `utils.hpp`
   * for all functions currently defined at the bottom of [`src/core/reaction_ensemble.hpp`](https://github.com/espressomd/espresso/blob/415f8d4f66bd8b1bb417a7617c14b01577834dad/src/core/reaction_ensemble.hpp#L397-L454)
* `ReactionAlgorithm.hpp`
   * `class ReactionAlgorithm`
   * `class SingleReaction`
   * `class StoredParticleProperty`
* `WangLandauReactionEnsemble.hpp`
   * `class WangLandauReactionEnsemble`
   * `class CollectiveVariable`
   * `class EnergyCollectiveVariable`
   * `class DegreeOfAssociationCollectiveVariable`

### PR DESCRIPTION
**Pull request by jngrad**
Fixes #4155

Description of changes:
- move each class in the ReactionEnsemble into a separate source file
   - helps with separation of concerns and readability
   - Wang-Landau and its collective variables are in the same file due to coupling
- rename the namespace `ReactionEnsemble` to `ReactionMethods`
   - fixes the name collision with class `ReactionEnsemble` (used to be an issue in the class unit test)
   - chose `ReactionMethods` over the other candidates proposed in #3284 since the main class `ReactionAlgorithm` is currently still designed to use the `SingleReaction` class; the discussion in #3284 regarding class names remains open
- move the classes and unit tests to subfolder `src/core/reaction_methods`

### CODE REVIEW

File: src/core/reaction_methods/ReactionAlgorithm.cpp [hunk 1]
```diff
@@ -0,0 +1,651 @@
...
+  init_type_map(non_interacting_type);
+
+  reactions.push_back(new_reaction);
+}
+
+/**
+ * Checks whether all necessary variables for the reaction ensemble have been
+ * set.
+ */
+void ReactionAlgorithm::check_reaction_ensemble() const {
```
Comment (jonaslandsgesell) [inline src/core/reaction_methods/ReactionAlgorithm.cpp]: This function could be renamed to check_reaction_method()

File: src/core/reaction_methods/ReactionAlgorithm.cpp [hunk 2]
```diff
@@ -0,0 +1,651 @@
...
+        throw std::runtime_error(message);
+      }
+    }
+  }
+#endif
+}
+
+/**
+ * Automatically sets the volume which is used by the reaction ensemble to the
+ * volume of a cuboid box
```
Comment (jonaslandsgesell) [inline src/core/reaction_methods/ReactionAlgorithm.cpp]: If it was not initialized with some other value

File: src/core/reaction_methods/tests/ConstantpHEnsemble_test.cpp [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+  BOOST_CHECK_THROW(r_algo.check_reaction_ensemble(), std::runtime_error);
+
+  // create a reaction A -> B + C
+  int const type_A = 0;
+  int const type_B = 1;
+  int const type_C = 2;
+  SingleReaction const reaction(2., {type_A}, {1}, {type_B, type_C}, {1, 1});
+
+  // check acceptance probability
+  constexpr auto g = factorial_Ni0_divided_by_factorial_Ni0_plus_nu_i;
```
Comment (jonaslandsgesell) [inline src/core/reaction_methods/tests/ConstantpHEnsemble_test.cpp]: This is unused in the constant pH method, consider removal of g

File: src/core/reaction_methods/tests/ConstantpHEnsemble_test.cpp [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+  for (int i = 0; i < 3; ++i) {
+    for (int j = 0; j < 3; ++j) {
+      for (int k = 0; k < 3; ++k) {
+        // system contains i x A, j x B, and k x C
+        auto const p_numbers =
+            std::map<int, int>{{type_A, i}, {type_B, j}, {type_C, k}};
+        auto const energy = static_cast<double>(i + 1);
+        auto const acceptance_ref =
```
Comment (jonaslandsgesell) [inline src/core/reaction_methods/tests/ConstantpHEnsemble_test.cpp:75]: Maybe a comment with the implemented formula is helpful?

File: src/core/reaction_methods/tests/ReactionEnsemble_test.cpp [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+  constexpr auto g = factorial_Ni0_divided_by_factorial_Ni0_plus_nu_i;
+  for (int i = 0; i < 3; ++i) {
+    for (int j = 0; j < 3; ++j) {
+      for (int k = 0; k < 3; ++k) {
+        // system contains i x A, j x B, and k x C
+        auto const p_numbers =
+            std::map<int, int>{{type_A, i}, {type_B, j}, {type_C, k}};
+        auto const energy = static_cast<double>(i + 1);
+        auto const f_expr = calculate_factorial_expression(reaction, p_numbers);
+        auto const acceptance_ref = 2e6 * f_expr * std::exp(energy / 20.);
```
Comment (jonaslandsgesell) [inline src/core/reaction_methods/tests/ReactionEnsemble_test.cpp]: Maybe a comment with the implemented formula is helpful?  20 is r_algo.temperature?

File: src/core/reaction_methods/tests/WangLandauReactionEnsemble_test.cpp [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+  constexpr auto g = factorial_Ni0_divided_by_factorial_Ni0_plus_nu_i;
+  for (int i = 0; i < 3; ++i) {
+    for (int j = 0; j < 3; ++j) {
+      for (int k = 0; k < 3; ++k) {
+        // system contains i x A, j x B, and k x C
+        auto const p_numbers =
+            std::map<int, int>{{type_A, i}, {type_B, j}, {type_C, k}};
+        auto const energy = static_cast<double>(i + 1);
+        auto const f_expr = calculate_factorial_expression(reaction, p_numbers);
+        auto const prefactor = 2e6 * f_expr;
```
Comment (jonaslandsgesell) [inline src/core/reaction_methods/tests/WangLandauReactionEnsemble_test.cpp:124]: Can you add a comment on the magic number 2e6?

File: src/core/reaction_methods/tests/ConstantpHEnsemble_test.cpp [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+            std::exp(energy / 20. + std::log(10.) * (1. + std::log10(2.)));
```
Comment (jonaslandsgesell) [inline src/core/reaction_methods/tests/ConstantpHEnsemble_test.cpp]: 20 is r_algo.temperature?

**Review by jonaslandsgesell — COMMENTED** [review]
Minor Comments

File: src/core/reaction_methods/tests/ConstantpHEnsemble_test.cpp (hunk 3 above)
Comment (jngrad) [reply src/core/reaction_methods/tests/ConstantpHEnsemble_test.cpp]: good find!

File: src/core/reaction_methods/ReactionAlgorithm.cpp (hunk 1 above)
Comment (jngrad) [reply src/core/reaction_methods/ReactionAlgorithm.cpp]: done

File: src/core/reaction_methods/ReactionAlgorithm.cpp (hunk 2 above)
Comment (jngrad) [reply src/core/reaction_methods/ReactionAlgorithm.cpp]: done

**Comment by jonaslandsgesell** [comment]
Looks good to me!

**Review by jonaslandsgesell — APPROVED** [review]

**Review by KaiSzuttor — APPROVED** [review]

**Pull request merged**